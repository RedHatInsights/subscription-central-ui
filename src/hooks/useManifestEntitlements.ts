import { useQuery, QueryObserverResult } from 'react-query';
import Cookies from 'js-cookie';

interface ManifestEntitlement {
  id: string;
  sku: string;
  contractNumber: string;
  entitlementQuantity: number;
  startDate?: string;
  endDate?: string;
  subscriptionName?: string;
}

interface ManifestEntitlementsData {
  body: {
    simpleContentAccess: string;
    createdBy: string;
    createdDate: string;
    entitlementsAttached: EntitlementsAttachedData;
    entitlementsAttachedQuantity: number;
    lastModified: string;
    name: string;
    type: string;
    uuid: string;
    version: string;
  };
}

interface EntitlementsAttachedData {
  valid: boolean;
  reason?: string;
  value?: ManifestEntitlement[];
}

const getManifestEntitlements = (uuid: string): Promise<ManifestEntitlementsData> => {
  const jwtToken = Cookies.get('cs_jwt');
  return fetch(`/api/rhsm/v2/manifests/${uuid}?include=entitlements`, {
    headers: { Authorization: `Bearer ${jwtToken}` }
  }).then((response) => response.json());
};

const useManifestEntitlements = (
  uuid: string
): QueryObserverResult<ManifestEntitlementsData, unknown> => {
  return useQuery<any, Error>(['manifestEntitlements', uuid], () => getManifestEntitlements(uuid));
};

export {
  EntitlementsAttachedData,
  ManifestEntitlementsData,
  ManifestEntitlement,
  useManifestEntitlements as default
};
