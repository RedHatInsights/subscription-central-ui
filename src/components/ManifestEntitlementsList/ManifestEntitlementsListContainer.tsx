import React, { FC } from 'react';
import ManifestEntitlementsList from './ManifestEntitlementsList';
import useManifestEntitlements from '../../hooks/useManifestEntitlements';

interface ManifestEntitlementsListContainerProps {
  uuid: string;
  entitlementsRowRef: React.MutableRefObject<HTMLSpanElement | HTMLParagraphElement>;
}
const ManifestEntitlementsListContainer: FC<ManifestEntitlementsListContainerProps> = ({
  uuid,
  entitlementsRowRef
}) => {
  const { isLoading, isSuccess, isError, data } = useManifestEntitlements(uuid);

  return (
    <ManifestEntitlementsList
      isLoading={isLoading}
      isSuccess={isSuccess}
      isError={isError}
      entitlementsData={data?.body?.entitlementsAttached}
      entitlementsRowRef={entitlementsRowRef}
    />
  );
};

export default ManifestEntitlementsListContainer;
