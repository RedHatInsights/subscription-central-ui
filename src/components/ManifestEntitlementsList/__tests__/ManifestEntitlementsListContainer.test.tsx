import React from 'react';
import ManifestEntitlementsListContainer from '../ManifestEntitlementsListContainer';
import { QueryClient, QueryClientProvider } from 'react-query';
import { render, waitFor } from '@testing-library/react';
import useManifestEntitlements from '../../../hooks/useManifestEntitlements';

jest.mock('../../../hooks/useManifestEntitlements');

const queryClient = new QueryClient();

describe('Manifest Entitlements List Container', () => {
  it('renders correctly once entitlements have been loaded', async () => {
    (useManifestEntitlements as jest.Mock).mockReturnValue({
      isLoading: false,
      isSuccess: true,
      isError: false,
      data: {
        body: {
          entitlementsAttached: {
            valid: true,
            value: [
              {
                contractNumber: '12345',
                entitlementQuantity: 10,
                id: 'id123',
                sku: 'sku123',
                startDate: '2021-01-01T00:00:00.000Z',
                endDate: '2022-01-01T00:00:00.000Z'
              }
            ]
          }
        }
      }
    });
    const props = {
      uuid: 'abc123',
      entitlementsRowRef: null as React.MutableRefObject<HTMLSpanElement>
    };
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        <ManifestEntitlementsListContainer {...props} />
      </QueryClientProvider>
    );

    await waitFor(() => expect(useManifestEntitlements).toHaveBeenCalledTimes(1));

    expect(container).toMatchSnapshot();
  });
});
