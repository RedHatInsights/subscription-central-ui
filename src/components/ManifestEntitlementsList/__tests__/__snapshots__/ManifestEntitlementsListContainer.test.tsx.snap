// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manifest Entitlements List Container renders correctly once entitlements have been loaded 1`] = `
<div>
  <div>
    <table
      aria-label="Allocations table"
      class="sub-c-table-manifests-entitlement-list pf-c-table pf-m-grid-md pf-m-no-border-rows"
      data-ouia-component-id="OUIA-Generated-Table-2"
      data-ouia-component-type="PF4/Table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class=""
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-1"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <th
            class="pf-m-nowrap"
            data-key="0"
            data-label="Subscription name"
            scope="col"
          >
            Subscription name
          </th>
          <th
            class="pf-m-nowrap"
            data-key="1"
            data-label="SKU"
            scope="col"
          >
            SKU
          </th>
          <th
            class="pf-m-nowrap"
            data-key="2"
            data-label="Contract number"
            scope="col"
          >
            Contract number
          </th>
          <th
            class="pf-m-nowrap"
            data-key="3"
            data-label="Quantity"
            scope="col"
          >
            Quantity
          </th>
          <th
            class="pf-m-nowrap"
            data-key="4"
            data-label="Start date"
            scope="col"
          >
            Start date
          </th>
          <th
            class="pf-m-nowrap"
            data-key="5"
            data-label="End date"
            scope="col"
          >
            End date
          </th>
        </tr>
      </thead>
      <tbody
        class=""
        role="rowgroup"
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-2"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Subscription name"
          />
          <td
            class=""
            data-key="1"
            data-label="SKU"
          >
            sku123
          </td>
          <td
            class=""
            data-key="2"
            data-label="Contract number"
          >
            12345
          </td>
          <td
            class=""
            data-key="3"
            data-label="Quantity"
          >
            10
          </td>
          <td
            class=""
            data-key="4"
            data-label="Start date"
          >
            2021-01-01
          </td>
          <td
            class=""
            data-key="5"
            data-label="End date"
          >
            2022-01-01
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
